### ISSUE TITLE
Regression: Callbacks stopped working due to #1795

### ISSUE THREAD

**Issue by maltelorbach** [ultralytics/ultralytics#1845]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

PR https://github.com/ultralytics/ultralytics/pull/1795 broke the callbacks. Callbacks are not passed on to the trainer, validator and exporter classes and thus are not called anymore.

Callbacks used to be registerd by adding them to `callbacks.default_callbacks`. Now they are added to a local variable `self.callbacks`. The trainer, validator and export classes have no access to that variable and only call the default callbacks.

This seems to be fixed for the predictor class, though, by passing the callbacks via argument `_callbacks`. This should be implemented for the other classes as well.

### Minimal Reproducible Example

This code below should stop the training immediately once "on_train_start" callback is called. Instead training continues running.
```
import sys

from ultralytics import YOLO


def printme(trainer):
    print("CALLBACK PRINTME CALLED")
    sys.exit(1)


def train():
    model = YOLO()
    model.add_callback("on_train_start", printme)
    model.train(epochs=1, data="coco128.yaml")

train()
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by dh031200** [comment]
@maltelorbach 
Add _callbacks arguments for every tasks.

And I check with the code below:
```python
import sys

from ultralytics import YOLO


def printme():
    print("CALLBACK PRINTME CALLED")
    sys.exit(1)


def train():
    model = YOLO()
    model.add_callback("on_train_start", printme)
    model.train(epochs=1, data="coco128.yaml")

def valid():
    model = YOLO()
    model.add_callback("on_val_start", printme)
    model.val()

# 1. Train
# train()

# 2. Valid
valid()
```

I made sure both callbacks are executed and terminated.

**Comment by glenn-jocher** [comment]
@maltelorbach @dh031200 should we revert https://github.com/ultralytics/ultralytics/pull/1795 and start from scratch here? I'm not sure if we are patching one problem and creating another again without more comprehensive callbacks CI/tests.

@AyushExel can you create some callbacks tests and take a closer look at the changes being created by PR's #1795 and #1846?

**Comment by glenn-jocher** [comment]
@maltelorbach @dh031200 in summary #1795 should have never been allowed to create a problem if we had callbacks properly tested in tests and/or CI.

**Comment by glenn-jocher** [comment]
@maltelorbach @dh031200 I've reverted https://github.com/ultralytics/ultralytics/pull/1795 with https://github.com/ultralytics/ultralytics/pull/1847 now and published as 8.0.64. 

We'll need to put more tests in place surrounding callbacks, and @AyushExel like you mentioned add single and multi GPU runners as a priority.
